### PR TITLE
Refactor WaveRNN infer and move to codebase

### PR DESCRIPTION
**Pull request by yangarbiter**
As discussed in #1651 

The training part of the WaveRNN seems to support two loss functions -- CrossEntropy and MolLoss
(https://github.com/pytorch/audio/blob/main/examples/pipeline_wavernn/losses.py). Currently the inference code only supports the model trained with CrossEntropy since there is no pretrained model trained with MolLoss to test it out yet.

### CODE REVIEW

File: torchaudio/models/wavernn.py [hunk 1]
```diff
@@ -347,6 +355,68 @@ def forward(self, waveform: Tensor, specgram: Tensor) -> Tensor:
...
+
+            logits = self.fc3(x)
+
+            posterior = F.softmax(logits, dim=1)
+
+            x = torch.multinomial(posterior, 1).float()
+            x = bits_to_normalized_waveform(x, n_bits)
+            output.append(x)
+
+        return torch.stack(output).transpose(0, 1).transpose(1, 2)
```
Comment (nateanl) [inline torchaudio/models/wavernn.py:418]: nit: is (1, 2, 0) the final dimension for the output?
```suggestion
        return torch.stack(output).permute(1, 2, 0)
```

File: torchaudio/models/wavernn.py [hunk 2]
```diff
@@ -347,6 +355,68 @@ def forward(self, waveform: Tensor, specgram: Tensor) -> Tensor:
         # bring back channel dimension
         return x.unsqueeze(1)
 
+    @torch.jit.export
+    def infer(self, specgram: Tensor) -> Tensor:
+        r"""Inference method of WaveRNN.
+
+        This function currently only supports multinomial sampling, which assumes the
+        network is trained on crossentropy loss.
```
Comment (nateanl) [inline torchaudio/models/wavernn.py:363]:
```suggestion
        network is trained on cross_entropy loss.
```

File: examples/pipeline_wavernn/wavernn_inference_wrapper.py [hunk 3]
```diff
@@ -166,8 +160,8 @@ def forward(self,
                 loss_name: str = "crossentropy",
                 mulaw: bool = True,
                 batched: bool = True,
-                timesteps: int = 11000,
-                overlap: int = 550) -> Tensor:
+                timesteps: int = 100,
+                overlap: int = 5) -> Tensor:
```
Comment (nateanl) [inline examples/pipeline_wavernn/wavernn_inference_wrapper.py:164]: nit: update the default values in the doc.

Also could you update the ``shape`` term in the docs? So that we can keep it in consistent with CONTRIBUTING.md.

File: torchaudio/models/wavernn.py [hunk 4]
```diff
@@ -347,6 +355,68 @@ def forward(self, waveform: Tensor, specgram: Tensor) -> Tensor:
...
+
+        Args:
+            specgram (Tensor): The input spectrogram to the WaveRNN with shape (n_batch, n_freq, n_time).
```
Comment (nateanl) [inline torchaudio/models/wavernn.py:366]:
```suggestion
            specgram (Tensor): The input spectrogram to the WaveRNN of size (n_batch, n_freq, n_time).
```

File: torchaudio/models/wavernn.py [hunk 5]
```diff
@@ -347,6 +355,68 @@ def forward(self, waveform: Tensor, specgram: Tensor) -> Tensor:
...
+
+        Return:
+            waveform (Tensor): The inferred waveform with shape (n_batch, 1, n_time).
```
Comment (nateanl) [inline torchaudio/models/wavernn.py:369]:
```suggestion
            waveform (Tensor): The inferred waveform of size (n_batch, 1, n_time).
```

File: examples/pipeline_wavernn/wavernn_inference_wrapper.py [hunk 6]
```diff
@@ -21,18 +21,12 @@
 # *****************************************************************************
 
 
-from typing import List
-
 from torchaudio.models.wavernn import WaveRNN
 import torch
-import torch.nn.functional as F
 import torchaudio
```
Comment (nateanl) [inline examples/pipeline_wavernn/wavernn_inference_wrapper.py:26]: Not sure if it follows Google's style check. I need your opinion on this to help me understand it.

```
import torch
from torch import Tensor
import torchaudio
from torchaudio.models.wavernn import WaveRNN
```
Or it's like 
```
import torch
import torchaudio
from torch import Tensor
from torchaudio.models.wavernn import WaveRNN
```

**Review by nateanl — APPROVED** [review]
LGTM overall. Just need to address the nits then we can merge it.

File: examples/pipeline_wavernn/inference.py [hunk 7]
```diff
@@ -39,11 +39,11 @@ def parse_args():
         help="Select the WaveRNN checkpoint."
     )
     parser.add_argument(
-        "--batch-timesteps", default=11000, type=int,
+        "--batch-timesteps", default=100, type=int,
```
Comment (mthrok) [inline examples/pipeline_wavernn/inference.py:42]: What are the impacts of these changes?

File: torchaudio/models/wavernn.py [hunk 8]
```diff
@@ -35,6 +36,13 @@
 }
 
 
+def bits_to_normalized_waveform(label: torch.Tensor, bits: int) -> torch.Tensor:
```
Comment (mthrok) [inline torchaudio/models/wavernn.py:39]: Should this be public?

File: examples/pipeline_wavernn/inference.py (hunk 7 above)
Comment (yangarbiter) [reply examples/pipeline_wavernn/inference.py:42]: When we want to use batch prediction, we would like to fold the signal on time domain first and then do the batch inference. This argument affects the folding length of the signal. For example:

original signal: [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]]
if this value set to 5, it would be 
[[0, 1, 2, 3, 4]
 [5, 6, 7, 8, 9]]
if this value set to 3, it would be 
[[0, 1, 2],
 [3, 4, 5],
 [7, 8, 9]]

We need to make this change because in the original implementation, the upsampling is applied first and then we perform the folding (see [here](https://github.com/pytorch/audio/pull/1704/files#diff-9b05c2c6426592ffa5157949e3e853bb3eae33276b34e837e0bae9751e10adcdL191-L200)). However, to move the folding of the signal outside of the infer method, we would need to fold first and then do the upsampling. To fold before the upsampling, the number of signals will be a lot lesser.

I think this change of the implementation will affect the output waveform a little, but I couldn't hear the difference before and after the change.

File: examples/pipeline_wavernn/wavernn_inference_wrapper.py (hunk 3 above)
Comment (yangarbiter) [reply examples/pipeline_wavernn/wavernn_inference_wrapper.py:164]: Thanks for the reminder, I've changed all `with shape` to `of size`.

File: torchaudio/models/wavernn.py (hunk 8 above)
Comment (yangarbiter) [reply torchaudio/models/wavernn.py]: No, it shouldn't. I've changed it to a private function [here](https://github.com/pytorch/audio/pull/1704/files#diff-7f4ad259da182d3840fe4c75dd0c0b7b9fe30f31503df0eff87b1e5c9378cac4R39-R43).

File: examples/pipeline_wavernn/wavernn_inference_wrapper.py (hunk 3 above)
Comment (yangarbiter) [reply examples/pipeline_wavernn/wavernn_inference_wrapper.py:164]: Thanks for the reminder, I've changed all `with shape` to `of size`.

**Comment by yangarbiter** [comment]
I've also made an additional change [here](https://github.com/pytorch/audio/pull/1704/files#diff-7f4ad259da182d3840fe4c75dd0c0b7b9fe30f31503df0eff87b1e5c9378cac4R359) so that in the future, we can easily support difference sampling mode. We may want this interface because the `pipeline_wavernn` supports the MolLoss and if a model is trained with that loss, it requires a different sampling method.

File: torchaudio/models/wavernn.py (hunk 4 above)
Comment (yangarbiter) [reply torchaudio/models/wavernn.py]: Fixed

File: torchaudio/models/wavernn.py (hunk 5 above)
Comment (yangarbiter) [reply torchaudio/models/wavernn.py]: Fixed

File: examples/pipeline_wavernn/wavernn_inference_wrapper.py (hunk 6 above)
Comment (yangarbiter) [reply examples/pipeline_wavernn/wavernn_inference_wrapper.py:26]: This part I am not sure since it doesn't trigger flake8's warning.
@mthrok what is our style for sorting the imports?

File: torchaudio/models/wavernn.py (hunk 1 above)
Comment (yangarbiter) [reply torchaudio/models/wavernn.py]: Thanks for the suggestion, I've fixed it [here](https://github.com/pytorch/audio/pull/1704/files#diff-7f4ad259da182d3840fe4c75dd0c0b7b9fe30f31503df0eff87b1e5c9378cac4R425).

File: torchaudio/models/wavernn.py (hunk 2 above)
Comment (yangarbiter) [reply torchaudio/models/wavernn.py]: Thanks for the comment, I've fixed it [here](https://github.com/pytorch/audio/pull/1704/files#diff-7f4ad259da182d3840fe4c75dd0c0b7b9fe30f31503df0eff87b1e5c9378cac4R363)

File: examples/pipeline_wavernn/wavernn_inference_wrapper.py (hunk 6 above)
Comment (mthrok) [reply examples/pipeline_wavernn/wavernn_inference_wrapper.py:26]: Currently, we do not have a style for sorting imports for historical reason.
So, we resorted to just make sure flake8 does not complain.

Probably it's a good time to start linting as a part of https://fburl.com/jjsby41j

File: torchaudio/models/wavernn.py [hunk 9]
```diff
@@ -35,6 +36,13 @@
 }
 
 
+def _bits_to_normalized_waveform(label: torch.Tensor, bits: int) -> torch.Tensor:
```
Comment (mthrok) [inline torchaudio/models/wavernn.py:39]: Since this function is used by only one place, why don't we hardcode it in the call site, with the docsrtings?

**Review by mthrok — APPROVED** [review]

File: torchaudio/models/wavernn.py (hunk 9 above)
Comment (yangarbiter) [reply torchaudio/models/wavernn.py]: Thanks for the suggestion, I've fixed it [here](https://github.com/pytorch/audio/pull/1704/files#diff-7f4ad259da182d3840fe4c75dd0c0b7b9fe30f31503df0eff87b1e5c9378cac4R412)

File: torchaudio/models/wavernn.py [hunk 10]
```diff
@@ -347,6 +348,76 @@ def forward(self, waveform: Tensor, specgram: Tensor) -> Tensor:
...
+            x = x + h2[0]
+            x = torch.cat([x, a3_t], dim=1)
+            x = F.relu(self.fc1(x))
+
+            x = torch.cat([x, a4_t], dim=1)
+            x = F.relu(self.fc2(x))
+
+            logits = self.fc3(x)
+
+            if sampling_mode == "multinomial":
```
Comment (mthrok) [inline torchaudio/models/wavernn.py:407]: Since, there is only one `sampling_mode`, we can get rid of it.

Comment (yangarbiter) [reply torchaudio/models/wavernn.py]: I've removed the sampling mode option and opened an issue to keep track of the support for MOLLoss (#1721).
